Add about rmm modes in `cudf.pandas` docs (#16404)
rapidsai/cudf: docs/cudf/source/cudf_pandas/how-it-works.md

When using `cudf.pandas`, cuDF's [pandas compatibility
mode](api.options) is automatically enabled, ensuring consistency with
pandas-specific semantics like default sort ordering.

> `cudf.pandas` uses a managed pool memory by default, that will also enable

Reviewer: ```suggestion
`cudf.pandas` uses a managed memory pool by default. This allows `cudf.pandas` to process datasets larger than the memory of the GPU(s) it is running on. Managed memory prefetching is also enabled by default to improve memory access performance. For more information on CUDA Unified Memory (managed memory), performance and prefetching, see [this NVIDIA Developer blog post](https://developer.nvidia.com/blog/improving-gpu-memory-oversubscription-performance/).
```

> prefetching <<LINK to prefetch tutorial/demo>>.

Author: @vyasr Would you be able to add a tutorial link/demo about prefetch that might be appropriate for cudf users?

Reviewer 2: I'm not sure. What kind of information would you want to convey? How prefetching works from the perspective of CUDA? How a Python users would trigger it (in this case, they wouldn't do so themselves)?

Author: I was thinking more like this: https://resources.nvidia.com/en-us-nsight-developer-tools/boosting-application?lx=P1ZhhI&topic=hpc&linkId=100000269652539&ncid=no-ncid

Reviewer: I think explaining prefetching requires first explaining what CUDA Unified Memory is, what virtual memory paging is, what a page fault is and why we want to avoid them, and then how prefetching helps avoid them.  That's a lot. I think we should just link to the best post or documentation we have that covers all of that. 

This one seems to fit the bill: https://developer.nvidia.com/blog/improving-gpu-memory-oversubscription-performance/

Author: Thanks @harrism ! That is a perfect !!

> There are various memory allocators that can be used by changing the environment
> variable `CUDF_PANDAS_RMM_MODE`. It supports:
> 
> 1. "pool"
> 2. "async"
> 3. "managed" (default)
> 4. "managed_pool"
> 5. "cuda"

Reviewer: I believe `managed_pool` is the default, not `managed`? Also suggesting more detail, and reordered (sort of) in order of preferred use (?).

```suggestion
Other memory allocators can be used by changing the environment
variable `CUDF_PANDAS_RMM_MODE` to one of the following.

Pool allocators improve allocation performance. Without using one, memory 
allocation may be a bottleneck depending on the workload. Managed memory
enables oversubscribing GPU memory. This allows cudf.pandas to process 
data larger than GPU memory in many cases, without CPU (Pandas) fallback.

1. "managed_pool" (default): CUDA Unified Memory (managed memory) with RMM's asynchronous pool allocator.
2. "managed": CUDA Unified Memory, (managed memory) with no pool allocator.
3. "async": CUDA's built-in pool asynchronous pool allocator with normal CUDA device memory. 
4. "pool": RMM's asynchronous pool allocator with normal CUDA device memory.
5. "cuda": normal CUDA device memory with no pool allocator.
```

When using `cudf.pandas`, cuDF's [pandas compatibility
mode](api.options) is automatically enabled, ensuring consistency with
pandas-specific semantics like default sort ordering.

> 

Reviewer 3: ```suggestion
```

> `cudf.pandas` uses a managed memory pool by default. This allows `cudf.pandas` to process datasets larger than the memory of the GPU(s) it is running on. Managed memory prefetching is also enabled by default to improve memory access performance. For more information on CUDA Unified Memory (managed memory), performance and prefetching, see [this NVIDIA Developer blog post](https://developer.nvidia.com/blog/improving-gpu-memory-oversubscription-performance/).

Reviewer 3: We don’t have a good multi-GPU solution right now so we should make GPU singular.
```suggestion
`cudf.pandas` uses a managed memory pool by default. This allows `cudf.pandas` to process datasets larger than the memory of the GPU it is running on. Managed memory prefetching is also enabled by default to improve memory access performance. For more information on CUDA Unified Memory (managed memory), performance, and prefetching, see [this NVIDIA Developer blog post](https://developer.nvidia.com/blog/improving-gpu-memory-oversubscription-performance/).
```

When using `cudf.pandas`, cuDF's [pandas compatibility
mode](api.options) is automatically enabled, ensuring consistency with
pandas-specific semantics like default sort ordering.

`cudf.pandas` uses a managed memory pool by default. This allows `cudf.pandas` to process datasets larger than the memory of the GPU(s) it is running on. Managed memory prefetching is also enabled by default to improve memory access performance. For more information on CUDA Unified Memory (managed memory), performance and prefetching, see [this NVIDIA Developer blog post](https://developer.nvidia.com/blog/improving-gpu-memory-oversubscription-performance/).

> prefetching <<LINK to prefetch tutorial/demo>>.

Reviewer 3: @galipremsagar We need to address this before admin merging.

Reviewer 3: ```suggestion
```

pandas-specific semantics like default sort ordering.

`cudf.pandas` uses a managed memory pool by default. This allows `cudf.pandas` to process datasets larger than the memory of the GPU(s) it is running on. Managed memory prefetching is also enabled by default to improve memory access performance. For more information on CUDA Unified Memory (managed memory), performance and prefetching, see [this NVIDIA Developer blog post](https://developer.nvidia.com/blog/improving-gpu-memory-oversubscription-performance/).
prefetching <<LINK to prefetch tutorial/demo>>.

Other memory allocators can be used by changing the environment
variable `CUDF_PANDAS_RMM_MODE` to one of the following.

Pool allocators improve allocation performance. Without using one, memory 
allocation may be a bottleneck depending on the workload. Managed memory
enables oversubscribing GPU memory. This allows cudf.pandas to process 

> data larger than GPU memory in many cases, without CPU (Pandas) fallback.

Reviewer: Maybe this should go after the list since we have "to one of the following" at the end of the paragraph above.

Author: Done 👍
